---
title: "sRNAs"
author: "Eddy_Mendoza"
date: "2023-04-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(data.table)
library(reshape2)
library(FactoMineR)
library(UpSetR)
library(DESeq2)
library(edgeR)
#### Visualization
library(MetBrewer)
```

## 1 .- Exploration of the distribution of species abundance and lengths in the raw libraries

```{r species abundance and lengths}
# abundance of sRNA species
### Select only for 21,22 and 24
species_srna = function(path, sample_name, sex_id, tissue_id) {
  table = data.table(read.table(path))
  counts = as.data.frame(table[, .N, by='V2']) %>% mutate(sample = sample_name, sex = sex_id, tissue = tissue_id)
  abundance = as.data.frame(table[, .N, by='V1']) %>% mutate(sample = sample_name, sex = sex_id, tissue = tissue_id) %>% filter(N > 2)
  list(table, counts, abundance)
}

### FEMALE SAMPLES
f1l=species_srna("raw/sRNA_MGX/trimmed/F1L_final_trimming.fastq_count.tsv", "F1L", "Female", "Leaf")
f1b=species_srna("raw/sRNA_MGX/trimmed/F1B_final_trimming.fastq_count.tsv", "F1B", "Female", "Flower bud")

f2l=species_srna("raw/sRNA_MGX/trimmed/F2L_final_trimming.fastq_count.tsv", "F2L", "Female", "Leaf")
f2b=species_srna("raw/sRNA_MGX/trimmed/F2B_final_trimming.fastq_count.tsv", "F2B", "Female", "Flower bud")

f3l=species_srna("raw/sRNA_MGX/trimmed/F3L_final_trimming.fastq_count.tsv", "F3L", "Female", "Leaf")
f3b=species_srna("raw/sRNA_MGX/trimmed/F3B_final_trimming.fastq_count.tsv", "F3B", "Female", "Flower bud")

### MALE SAMPLES

m1l=species_srna("raw/sRNA_MGX/trimmed/M1L_final_trimming.fastq_count.tsv", "M1L", "Male", "Leaf")
m1b=species_srna("raw/sRNA_MGX/trimmed/M1B_final_trimming.fastq_count.tsv", "M1B", "Male", "Flower bud")

m2l=species_srna("raw/sRNA_MGX/trimmed/M2L_final_trimming.fastq_count.tsv", "M2L", "Male", "Leaf")
m2b=species_srna("raw/sRNA_MGX/trimmed/M2B_final_trimming.fastq_count.tsv", "M2B", "Male", "Flower bud")

m4l=species_srna("raw/sRNA_MGX/trimmed/M4L_final_trimming.fastq_count.tsv", "M4L", "Male", "Leaf")
m4b=species_srna("raw/sRNA_MGX/trimmed/M4B_final_trimming.fastq_count.tsv", "M4B", "Male", "Flower bud")

```

```{r abundance densities}
### Species abundance

spp = do.call(rbind, lapply(list(f1l[[3]],
             f1b[[3]],
             f2l[[3]],
             f2b[[3]],
             f3l[[3]],
             f3b[[3]],
             m1l[[3]],
             m1b[[3]],
             m2l[[3]],
             m2b[[3]],
             m4l[[3]],
             m4b[[3]]), as.data.frame
            ) )

rm(f1b, f1l, f2b, f2l, f3b, f3l, m1b, m1l, m2b, m2l, m4b, m4l)

colnames(spp) = c("sRNA", "counts", "sample", "sex", "tissue")

spp = spp %>% mutate(log_counts = log2(counts))

#### Histogram 

ggplot(spp) +
  geom_density(aes(x=log_counts, color=sample), alpha=0.6) + 
  facet_grid(tissue ~ sex) +
  scale_color_manual(values = met.brewer("Hiroshige", 12)) +
  labs(y="Density", x="log2(Expression)")

ggsave("results/sp_expression.png", width = 6, height = 5, dpi = 600)
```

```{r reduced }
### With 320880141 total reads and mincov of 1 reads per million, the min read depth is 321
### For RPM > 1
no_uniq = spp[spp$counts>321, ]
ggplot(no_uniq) +
  geom_density(aes(x=log2(counts), color=sample), alpha=0.6) + 
  facet_grid(tissue ~ sex) +
  scale_color_manual(values = met.brewer("Hiroshige", 12)) +
  labs(y="Density", x="log2(Expression)")

ggsave("results/reduced_sp_expression.png", width = 6, height = 5, dpi = 600)
```

```{r lengths}
lgt = do.call(rbind, lapply(list(f1l[[2]],
             f1b[[2]],
             f2l[[2]],
             f2b[[2]],
             f3l[[2]],
             f3b[[2]],
             m1l[[2]],
             m1b[[2]],
             m2l[[2]],
             m2b[[2]],
             m4l[[2]],
             m4b[[2]]), as.data.frame
            ) )

ggplot(lgt) +
  geom_bar(aes(x=as.factor(V2), y=as.numeric(N), fill=sample), stat="identity", position = "dodge") +
  facet_grid(tissue ~ sex) +
  scale_fill_manual(values = met.brewer("Hiroshige", 12)) +
  labs(x="sRNA Length", y="Count")

ggsave("results/length.png", width = 6, height = 5, dpi = 600)
```

After the alignment and quantification with ShortStack

```{r fraction of the aligned reads}
fracs = read.table("raw/ShortStack_results/alignment_details.tsv", header = T)
fracs = filter(fracs, mapping_type == "P", count > 0)  
fracs = cbind(data.frame(
  sample = rep(c(rep(1,6),rep(2,6), rep(3,6)), 2),
  sex = c(rep("Female", 18), rep("Male", 18)),
  organ = rep(c(rep("Flower bud", 3), rep("Leaf", 3)), 6)
) , fracs)
fracs = fracs %>% select(sample, sex, organ, read_length, count)

ggplot(fracs) + 
  geom_bar(aes(fill=read_length, y=count, x=sample), position="fill", stat="identity") +
  facet_grid(organ ~ sex) +
  scale_fill_manual(values = met.brewer("Hokusai3")) +
  labs(y="Percentage of aligned reads", x="Individual", fill="sRNA length") +
  theme(text = element_text(size=20))

ggsave("results/fractions.png", width = 6, height = 6, dpi = 600)
```

```{r pca}
### All together
exp_matrix = read.table("raw/ShortStack_results/Counts.txt", header = T)

transposed = t(exp_matrix[, 4:15])
colnames(transposed) = exp_matrix$Name
samples = data.frame(samples = rownames(transposed), 
                     tissue = c(rep(c("Female, Flower bud", "Female, Leaf"), 3), rep(c("Male, Flower bud", "Male, Leaf"), 3)),
                     sex = c(rep("Female", 6), rep("Male", 6)),
                     organ = rep(c("Flower bud", "Leaf"), 6)
                       )

pca_exp = PCA(transposed, ncp = 20, graph = F)

pcadf = data.frame(samples, 
                   pca1 = as.data.frame(pca_exp$ind)[,1],
                   pca2 = as.data.frame(pca_exp$ind)[,2],
                   pca3 = as.data.frame(pca_exp$ind)[,3])

ggplot(pcadf) +
  geom_hline(yintercept = 0, color="gray40")+
  geom_vline(xintercept = 0, color="gray40")+
  geom_jitter(aes(pca1, pca2, size=pca3, fill=tissue), shape=21) +
  labs(x="Component 1 (35.67%)", y= "Component 2 (13.39%)", size= "Component 3 (10.54%)", fill= "Tissue") +
  scale_fill_manual(values = c("#3ab98f", "#2ab1c6", "#ee4d80", "#b52ac6")) +
  scale_size(range = c(2,8)) +
  guides(fill = guide_legend(override.aes = list(size=8))) +
  coord_fixed(ratio = 1.5) +
  theme_bw(base_size = 20)

ggsave("results/pca.png", width = 10, height = 10, dpi = 600)
```

## 2.- Differential expression

```{r data prep, 21-22}
exp_ptgs = read.table("raw/only_21-22_microRNAs/Counts.txt", header = T)
# sRNA clusters names will be kept as row names, remove columns with ids
matrix_ptgs = exp_ptgs[, 4:15]
row.names(matrix_ptgs) = exp_ptgs$Name


# count the number of genes expressed in none of the samples
table(rowSums(matrix_ptgs)==0)

# A good threshold can be chosen by identifying the CPM (Count Per million calculated by the EdgeR package) 
# that corresponds to a count of 10
cpm(10, mean(colSums(matrix_ptgs))) #value = 0.7191507

# discard low counts entries for at least 2 samples
keep = rowSums(cpm(matrix_ptgs)>0.5) >= 2	
summary(keep) # 129 will be removed 
matrix_ptgs = matrix_ptgs[keep,]

# Prepare data info
datainfo = samples[, 2:4]
row.names(datainfo) = samples$samples # colnames in the expression matrix must match 
```

```{r DE function}

pyramide = function(raw_matrix, data_info, design_formula, groups_vector){

######### DESEQ 2

# Generate the DESeqDataSet
deseq_object = DESeqDataSetFromMatrix(countData = raw_matrix, colData = data_info, design = design_formula)

# Run Analysis
deseq_object = DESeq(deseq_object)

# Get table
deseq_table = as.data.frame(results(deseq_object, pAdjustMethod = "BH"))

# Get DE clusters that fall under the selection threshold (pvalue < 0.001)
deseq = rownames(deseq_table[which(deseq_table$pvalue < 0.001), ] )

######## EDGER

# generate DGEList with grouping by species 
edgeR.DGElist <- DGEList(counts = raw_matrix, group = groups_vector)

# normalize for RNA composition, scaling for the effective library size
edgeR.DGElist <- calcNormFactors(edgeR.DGElist)

# Estimation of  dispersion
edgeR.DGElist <- estimateCommonDisp(edgeR.DGElist, verbose=T)
edgeR.DGElist <- estimateTagwiseDisp(edgeR.DGElist)

# ESTIMATION OF DISPERTION by GLM
design.mat <- model.matrix(~ groups_vector)

# perform DE testing
fit <- glmFit(edgeR.DGElist, design.mat)
lrt <- glmLRT(fit , coef = 2)

# extract  results table
edger_table = topTags(lrt , n = Inf, sort.by = "PValue", adjust.method = "BH")$table

# Extract most differential expressed genes
edger = rownames(edger_table[which(edger_table$PValue < 0.001), ] )

######### LIMA VOOM 

# Redefine object
LimmaVoom.DGElist <- edgeR.DGElist

# normalize for RNA composition, scaling for the effective library size
LimmaVoom.DGElist <- calcNormFactors(LimmaVoom.DGElist)
rownames(design.mat) <- colnames(LimmaVoom.DGElist) # add names to the design matrix from edgeR

# Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and 
# use this to compute appropriate observational-level weights. The data are then ready for linear modelling
voomTransformed  <- voom(LimmaVoom.DGElist , design.mat , plot=F)

# fit a linear model for each gene
voomed.fitted  <- lmFit(voomTransformed , design = design.mat)

# compute  statistics
voomed.fitted  <- eBayes(voomed.fitted)

# extract most differential expressed genes and table
lvoom_table = topTable(voomed.fitted, adjust="BH", number = Inf, sort.by="P")
lvoom = rownames(lvoom_table[which(lvoom_table$P.Value < 0.001), ] )

######## Get the clusters that were diferentially expressed in at least 2 methods
clusters = unique(c(intersect(deseq, edger), intersect(deseq, lvoom), intersect(edger, lvoom)))

####### Print results
list("clusters" = clusters,
     "DeSeq2 table" = deseq_table,
     "EdgeR table" = edger_table,
     "Lima-Voom table" = lvoom_table)
}
```

### Differential Expression for sex-biased/specific siRNAs (21-22nt) in Flowers

```{r flowers ptgs}
# Select samples
fptgs = matrix_ptgs %>% select(grep('B', colnames(matrix_ptgs)))
fptgs_info = datainfo %>% filter(organ=="Flower bud")
# define groups (in the same order as the table)
fgroups = datainfo[datainfo$organ == "Flower bud", ]$sex

flowers = pyramide(fptgs, fptgs_info, ~sex, fgroups)
```


### Differential Expression for sex-biased/specific siRNAs (21-22nt) in Leaves

```{r leaves ptgs}
# Select samples
lptgs = matrix_ptgs %>% select(grep('L', colnames(matrix_ptgs)))
lptgs_info = datainfo %>% filter(organ=="Leaf")
# define groups (in the same order as the table)
lgroups = datainfo[datainfo$organ == "Leaf", ]$sex

leaves = pyramide(lptgs, lptgs_info, ~sex, lgroups)
```


### Differential Expression for tissue-biased/specific siRNAs (21-22nt) in Females

```{r female ptgs}
# Select samples
feptgs = matrix_ptgs %>% select(grep('F', colnames(matrix_ptgs)))
feptgs_info = datainfo %>% filter(sex=="Female")
# define groups (in the same order as the table)
fegroups = datainfo[datainfo$sex == "Female", ]$organ

female = pyramide(feptgs, feptgs_info, ~organ, fegroups)
```

### Differential Expression for tissue-biased/specific siRNAs (21-22nt) in Males

```{r male ptgs}
# Select samples
maptgs = matrix_ptgs %>% select(grep('M', colnames(matrix_ptgs)))
maptgs_info = datainfo %>% filter(sex=="Male")
# define groups (in the same order as the table)
magroups = datainfo[datainfo$sex == "Male", ]$organ

male = pyramide(maptgs, maptgs_info, ~organ, magroups)
```

### Intercepts for PTGS

```{r upset Deseq}
ptgs_list = list("Sex-biased in flower buds" = flowers[["clusters"]], 
                  "Sex-biased in leaves" = leaves[["clusters"]], 
                  "Tissue-biased in females" = female[["clusters"]], 
                  "Tissue-biased in males" = male[["clusters"]])

upset(fromList(ptgs_list))
```

```{r}
# Sex-biased in flowers and tissue-specific in males
baits_flm = intersect(flowers[["clusters"]], male[["clusters"]]) 
baits_flm = as.data.frame(cpm(matrix_ptgs, log=TRUE)[baits_flm, ] ) %>% select(grep('B', colnames(matrix_ptgs))) 
baits_flm = baits_flm %>% mutate(cluster = rownames(baits_flm)) %>% group_by(cluster) %>% summarise(female_exp = mean(c(F1B_dicer, F2B_dicer, F3B_dicer)), 
                                                                                                    male_exp = mean(c(M1B_dicer, M2B_dicer, M4B_dicer)), 
                                                                                                    sex_biased = male_exp>female_exp,
                                                                                                    organ = "Sex-biased in flowers",
                                                                                                    sex = "Tissue-specific in males") %>% add_count(organ)
# Sex-biased in flowers and tissue-specific in females
baits_flf = intersect(flowers[["clusters"]], female[["clusters"]]) 
baits_flf = as.data.frame(cpm(matrix_ptgs, log=TRUE)[baits_flf, ] ) %>% select(grep('B', colnames(matrix_ptgs))) 
baits_flf = baits_flf %>% mutate(cluster = rownames(baits_flf)) %>% group_by(cluster) %>% summarise(female_exp = mean(c(F1B_dicer, F2B_dicer, F3B_dicer)), 
                                                                                                    male_exp = mean(c(M1B_dicer, M2B_dicer, M4B_dicer)), 
                                                                                                    sex_biased = male_exp>female_exp,
                                                                                                    organ = "Sex-biased in flowers",
                                                                                                    sex = "Tissue-specific in females") %>% add_count(organ)
# Sex-biased in leaves and tissue-specific in males
baits_lvm = intersect(leaves[["clusters"]], male[["clusters"]]) 
baits_lvm = as.data.frame(cpm(matrix_ptgs, log=TRUE)[baits_lvm, ] ) %>% select(grep('L', colnames(matrix_ptgs))) 
baits_lvm = baits_lvm %>% mutate(cluster = rownames(baits_lvm)) %>% group_by(cluster) %>% summarise(female_exp = mean(c(F1L_dicer, F2L_dicer, F3L_dicer)), 
                                                                                                    male_exp = mean(c(M1L_dicer, M2L_dicer, M4L_dicer)), 
                                                                                                    sex_biased = male_exp>female_exp,
                                                                                                    organ = "Sex-biased in leaves",
                                                                                                    sex = "Tissue-specific in males") %>% add_count(organ)
# Sex-biased in leaves and tissue-specific in females
baits_lvf = intersect(leaves[["clusters"]], female[["clusters"]]) 
baits_lvf = as.data.frame(cpm(matrix_ptgs, log=TRUE)[baits_lvf, ] ) %>% select(grep('L', colnames(matrix_ptgs))) 
baits_lvf = baits_lvf %>% mutate(cluster = rownames(baits_lvf)) %>% group_by(cluster) %>% summarise(female_exp = mean(c(F1L_dicer, F2L_dicer, F3L_dicer)), 
                                                                                                    male_exp = mean(c(M1L_dicer, M2L_dicer, M4L_dicer)), 
                                                                                                    sex_biased = male_exp>female_exp,
                                                                                                    organ = "Sex-biased in leaves",
                                                                                                    sex = "Tissue-specific in females") %>% add_count(organ)
# Merge
bait_df = rbind(baits_flm, baits_flf, baits_lvm, baits_lvf)
bait_df$sex = factor(bait_df$sex, levels = c("Tissue-specific in males", "Tissue-specific in females"))

# Plot all
ggplot(bait_df) +
  annotate(geom = "polygon", x = c(-Inf, Inf, Inf), y = c(-Inf, Inf, -Inf), fill = "gray8", alpha = 0.2 ) +
  geom_abline(intercept = 0, slope = 1, alpha=0.6, color="red2", size=1, linetype="dashed") +
  geom_jitter(aes(female_exp, male_exp, fill=sex_biased), shape=21, size=6) +
  geom_text(aes(x = -Inf, y = -Inf, label= paste0("n = ", n)), hjust = -.5, vjust   = -28) +
  scale_x_continuous(limits = c(-2.5, 10)) +
  scale_y_continuous(limits = c(-2.5, 10)) +
  labs(x= "♀ logRPM", y="♂ logRPM") +
  scale_fill_manual(values = met.brewer("Derain")) +
  theme_bw() +
  theme(legend.position = "none",
        text = element_text(size = 20),
        panel.border = element_rect(color = "black", fill=NA),
        aspect.ratio=1) +
  facet_grid(sex~organ)

ggsave("results/desrnas.png", width = 8, height = 8, dpi = 600)
```

```{r}
f1b_y = read.table("raw/depth/F1B_sff.depth", header = F)
```

